submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,346 @@
+Service, it will have to reach the globalIP assigned to the exportedService and the response
+from the backend Pods will have the source as the globalIP of the exportedService.
+
+On the other hand, when the backend Pod tries to talk to the external application in remote
+Cluster, the external application will see that traffic is coming with a globalIP that is
+assigned either at the Cluster level, or namespace level, or the one requested by the user
+at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
+K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
+for both ingress and egress traffic can create Headless Services in front of them and should
+avoid requesting explicit globalIP that match Headless Service backend Pods.
+
+The proposal has its own set of Pros and Cons.
+
+### Pros
+
+Since we will be allocating globalIPs to only select K8s objects, it will help in improved scalability.
+Also, this translates to fewer iptable rules on the active Gateway node of the cluster which will
+improve the overall performance of the solution.
+
+### Cons
+
+A limitation with the proposed approach is that we will lose the unique identity for **every**
+Pod which is supported today. For users that intend to use Network Policies across Clusters or
+for debugging purposes, this can be mitigated by creating the necessary GlobalEgressIP CRDs
+which create unique globalIPs at the Namespace/Pod level.
+
+### Caveats
+
+In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIP,
+and the corresponding egress rules are programmed on the node, it was not able to communicate with
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Design Details
+
+The following CRD is proposed to support the new design.
+
+```Go
+type ClusterGlobalEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of ClusterGlobalEgressIP object.
+    Spec               ClusterGlobalEgressIPSpec `json:"spec"`
+
+    // Observed status of ClusterGlobalEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type ClusterGlobalEgressIPSpec struct {
+    // The number of globalIP's requested at the cluster-level.
+    // Globalnet Controller will allocate the requested number of contiguous GlobalIPs for
+    // this ClusterGlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.
+    // +optional
+    NumGlobalIPs  *int     'json:"numGlobalIPs",omitempty"`
```

Comment:
just a nit, we are shuttering with Global/Global. May be we should just use Size or NumIPs?

But a nit..

Reply:
I'm fine with it. We just merged the corresponding CRD in submariner, but it should not be a problem to update it.
@tpantelis what are your thoughts?

Reply:
NumIPs sounds fine.  Size seem a bit vague.